src/charm.py

```diff
@@ -214,12 +245,40 @@ def _is_tls_enabled(self) -> bool:
         """Return True if TLS is enabled."""
         return self.cert.enabled
 
+    def _is_tracing_enabled(self) -> bool:
+        """Return True if tracing is enabled."""
+        eps = self._tracing.endpoints
+        if not eps or not eps.host or not eps.ingesters:
+            return False
+        return True
+
+    def _on_tracing_endpoint_removed(self, event) -> None:
+        if not self.container.can_connect():
+            event.defer()
+            return
+        self._clear_tracing_config()
+
+    def _on_tracing_endpoint_changed(self, event) -> None:
+        # On slow machines, this event may come up before pebble is ready
+        if not self.container.can_connect():
+            event.defer()
```

Comment:
The logic captured in this defer will not be executed if we upgrade soon after deploying (and relation events do not re-fire on upgrade) so tracing config won't "ever" be written.

Reply:
on upgrade, we'll run `_clear_all_configs_and_restart_traefik` which in turn calls `_push_config` that in turn calls `_push_tracing_config`. So it's all right.

What I get from this is that my feeling that traefik's config management is getting too complex is confirmed. We should consider refactoring it to use manifests